### PR TITLE
Add beam search in transformer

### PR DESCRIPTION
**Pull request by zhangxiaoli73**
## What changes were proposed in this pull request?

Add beam search in transformer, details:

1. Enhance Attention layer to support more inputs in prediction
2. Enhance Transformer layer to support beam search in prediction

## How was this patch tested?

unit tests

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/XXX

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Attention.scala [hunk 1]
```diff
@@ -15,24 +15,58 @@
...
 /**
  * Implementation of multiheaded attention and self-attention layers.
+ *
  * @param hiddenSize hidden size
  * @param numHeads heads number
  * @param attentionDropout
  */
-class Attention[T: ClassTag](
-  val hiddenSize: Int, val numHeads: Int, val attentionDropout: Float)
- (implicit ev: TensorNumeric[T]) extends BaseModule[T] {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Attention.scala]: why not extending from `BaseModule`

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Attention.scala [hunk 2]
```diff
@@ -15,24 +15,58 @@
...
+class Attention[T: ClassTag](val hiddenSize: Int, val numHeads: Int, val attentionDropout: Float)
+  (implicit ev: TensorNumeric[T]) extends AbstractModule[Activity, Activity, T] {
+
+  // for prediction
+  private val join1 = nn.JoinTable[T](dimension = 2, nInputDims = -1)
+  private val join2 = nn.JoinTable[T](dimension = 2, nInputDims = -1)
+
+  private val queryLayer = TransformerOperation.dense(
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Attention.scala]: `join1` is not a good name, make it more meaningful ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/ThreadPool.scala [hunk 3]
```diff
@@ -107,7 +107,9 @@ class ThreadPool(private var poolSize: Int) {
 
       MKL.setNumThreads(size)
       BackendMklDnn.setNumThreads(size)
-      Affinity.setOmpAffinity()
+      if (System.getProperty("bigdl.disableOmpAffinity", "false").toBoolean) {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/ThreadPool.scala]: why this change ? thought already in master?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Attention.scala (hunk 1 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Attention.scala]: If extending from `BaseModule`, `buildModel` in parent class will be called before layer initialed in child class. Then error happens.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/ThreadPool.scala (hunk 3 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/ThreadPool.scala]: Yeah, I need to rebase master code.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Attention.scala (hunk 2 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Attention.scala]: ok, changed

**Comment by zhangxiaoli73** [comment]
jenkins pass: http://mlt-bddl01.sh.intel.com:8080/view/PR/job/BigDL-PRVN-lite/706/
